File: macos/Gutenkern/Sources/Gutenkern/L10n.swift
import Combine
import Foundation
import GutenkernCore

final class LanguageSettings: ObservableObject {
    static let shared = LanguageSettings()
    static let systemCode = "system"

    private static let defaultsKey = "interfaceLanguage"

    @Published var preference: String {
        didSet {
            UserDefaults.standard.set(preference, forKey: Self.defaultsKey)
        }
    }

    var resolvedLanguage: String {
        if preference != Self.systemCode, L10n.supported.contains(preference) {
            return preference
        }
        return L10n.systemLanguage
    }

    private init() {
        let stored = UserDefaults.standard.string(forKey: Self.defaultsKey) ?? Self.systemCode
        if stored == Self.systemCode || L10n.supported.contains(stored) {
            preference = stored
        } else {
            preference = Self.systemCode
        }
    }
}

struct AppLanguage {
    let code: String
    let nativeName: String
}

enum L10n {
    static let supported = ["en", "de", "fr", "pl", "cs", "es", "sv", "it", "uk"]
    static let languages: [AppLanguage] = [
        .init(code: "en", nativeName: "English"),
        .init(code: "de", nativeName: "Deutsch"),
        .init(code: "fr", nativeName: "Français"),
        .init(code: "pl", nativeName: "Polski"),
        .init(code: "cs", nativeName: "Čeština"),
        .init(code: "es", nativeName: "Español"),
        .init(code: "sv", nativeName: "Svenska"),
        .init(code: "it", nativeName: "Italiano"),
        .init(code: "uk", nativeName: "Українська")
    ]

    static var fieldWhat: String { t("fieldWhat") }
    static var groupCopied: String { t("groupCopied") }
    static var format: String { t("format") }
    static var formatFontLab: String { t("formatFontLab") }
    static var formatGlyphs: String { t("formatGlyphs") }
    static var result: String { t("result") }
    static var copy: String { t("copy") }
    static var save: String { t("save") }
    static var saveEllipsis: String { t("saveEllipsis") }
    static var saveAndOpenEllipsis: String { t("saveAndOpenEllipsis") }
    static var saveAsFile: String { t("saveAsFile") }
    static var saveFailed: String { t("saveFailed") }
    static var copyAll: String { t("copyAll") }
    static var strikeThrough: String { t("strikeThrough") }
    static var clearStrike: String { t("clearStrike") }
    static var resultPlaceholder: String { t("resultPlaceholder") }
    static var chooseFormat: String { t("chooseFormat") }
    static var resetProgress: String { t("resetProgress") }
    static var settings: String { t("settings") }
    static var language: String { t("language") }
    static var languageSystem: String { t("languageSystem") }
    static var about: String { t("about") }
    static var aboutBody: String { t("aboutBody") }
    static var aboutCredits: String { t("aboutCredits") }
    static var appName: String { t("appName") }
    static var arsenName: String { t("arsenName") }
    static var vladName: String { t("vladName") }
    static var oleksiiName: String { t("oleksiiName") }
    static let authorEmail = "armos1999@gmail.com"
    static let authorWebsite = "arsenmosiichuk.in.ua"
    static let vladWebsite = "zahrevsky.com"
    static let oleksiiWebsite = "oleksii.shmalko.com"
    static let updatesPath = "arsenmosiichuk.in.ua/gutenkern"
    static var authorEmailURL: URL { URL(string: "mailto:\(authorEmail)")! }
    static var authorWebsiteURL: URL { URL(string: "https://\(authorWebsite)")! }
    static var vladWebsiteURL: URL { URL(string: "https://\(vladWebsite)")! }
    static var oleksiiWebsiteURL: URL { URL(string: "https://\(oleksiiWebsite)")! }
    static var updatesURL: URL { URL(string: "https://\(updatesPath)")! }

    static func aboutBodyText() -> AttributedString {
        AttributedString(aboutBody.replacingOccurrences(of: "{appName}", with: appName))
    }

    static func aboutVersionText(_ version: String) -> AttributedString {
        let template = t("aboutVersion").replacingOccurrences(of: "{version}", with: version)
        return attributedTemplate(template, replacements: [
            "{updatesUrl}": (updatesPath, updatesURL)
        ])
    }

    static func aboutCreditsText() -> AttributedString {
        var text = attributedTemplate(aboutCredits, replacements: [
            "{arsenName}": (arsenName, authorWebsiteURL),
            "{vladName}": (vladName, vladWebsiteURL),
            "{oleksiiName}": (oleksiiName, oleksiiWebsiteURL)
        ])
        text.append(AttributedString(" "))
        var email = AttributedString(authorEmail)
        email.link = authorEmailURL
        text.append(email)
        text.append(AttributedString("."))
        return text
    }

    static func groupLabel(_ group: KerningGroup) -> String {
        groupName(group)
    }

    static func groupSidebarLabel(_ group: KerningGroup, done: Int, total: Int) -> String {
        var label = groupLabel(group)
        guard total > 0 else {
            return label
        }
        label += " ⋅ \(groupedCount(done))/\(groupedCount(total))"
        if done == total {
            label += " ✓"
        }
        return label
    }

    private static func groupName(_ group: KerningGroup) -> String {
        switch group {
        case .capitals: t("groupCapitals")
        case .smallCaps: t("groupSmallCaps")
        case .lowercase: t("groupLowercase")
        case .punctuation: t("groupPunctuation")
        case .nonAlphabetic: t("groupNonAlphabetic")
        case .liningFigures: t("groupLiningFigures")
        case .oldstyleFigures: t("groupOldstyleFigures")
        }
    }

    static let systemLanguage: String = resolveSystemLanguage()

    static func pairCount(_ count: Int) -> String {
        let key: String
        switch plural(count) {
        case .one: key = "pairOne"
        case .few: key = "pairFew"
        case .many: key = "pairMany"
        case .other: key = "pairOther"
        }
        return t(key).replacingOccurrences(of: "{count}", with: groupedCount(count))
    }

    static func pairProgress(done: Int, total: Int) -> String {
        let progress = "\(groupedCount(done))/\(groupedCount(total))"
        var label = pairCount(total).replacingOccurrences(of: groupedCount(total), with: progress)
        if total > 0, done == total {
            label += " ✓"
        }
        return label
    }

    static func groupedCount(_ count: Int) -> String {
        let sign = count < 0 ? "-" : ""
        let digits = Array(String(abs(count)))
        var grouped: [Character] = []
        for (index, digit) in digits.reversed().enumerated() {
            if index > 0, index % 3 == 0 {
                grouped.append("\u{202F}")
            }
            grouped.append(digit)
        }
        return sign + String(grouped.reversed())
    }

    private static func attributedTemplate(
        _ template: String,
        replacements: [String: (String, URL)]
    ) -> AttributedString {
        var result = AttributedString()
        var remaining = template[...]
        while let match = replacements.keys.compactMap({ key -> (key: String, index: String.Index)? in
            guard let index = remaining.range(of: key)?.lowerBound else {
                return nil
            }
            return (key, index)
        }).min(by: { $0.index < $1.index }) {
            let prefix = remaining[remaining.startIndex..<match.index]
            if !prefix.isEmpty {
                result.append(AttributedString(String(prefix)))
            }
            let (label, url) = replacements[match.key]!
            var link = AttributedString(label)
            link.link = url
            result.append(link)
            remaining = remaining[remaining.index(match.index, offsetBy: match.key.count)...]
        }
        if !remaining.isEmpty {
            result.append(AttributedString(String(remaining)))
        }
        return result
    }

    private static let catalog: [String: [String: String]] = loadCatalog()

    private static func t(_ key: String) -> String {
        table()[key] ?? key
    }

    private static func table() -> [String: String] {
        let language = LanguageSettings.shared.resolvedLanguage
        var merged = catalog["en"] ?? [:]
        if language != "en", let overlay = catalog[language] {
            merged.merge(overlay) { _, new in new }
        }
        return merged
    }

    private static func resolveSystemLanguage() -> String {
        let candidates = Locale.preferredLanguages + Bundle.main.preferredLocalizations
        for preferred in candidates {
            let code = languageCode(from: preferred)
            if supported.contains(code) {
                return code
            }
        }
        return "en"
    }

    private static func languageCode(from identifier: String) -> String {
        let normalized = identifier.replacingOccurrences(of: "_", with: "-")
        guard let first = normalized.split(separator: "-").first else { return "" }
        return String(first).lowercased()
    }

    private static func loadCatalog() -> [String: [String: String]] {
        guard
            let url = catalogURL(),
            let data = try? Data(contentsOf: url),
            let catalog = try? JSONDecoder().decode([String: [String: String]].self, from: data)
        else {
            return [:]
        }
        return catalog
    }

    private static func catalogURL() -> URL? {
        if let url = Bundle.main.url(forResource: "l10n", withExtension: "json") {
            return url
        }

        let fromSource = URL(fileURLWithPath: #file)
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .deletingLastPathComponent()
            .appendingPathComponent("core/l10n.json")
        if FileManager.default.fileExists(atPath: fromSource.path) {
            return fromSource
        }
        return nil
    }

    private enum Plural {
        case one, few, many, other
    }

    private static func plural(_ count: Int) -> Plural {
        let n = abs(count)
        switch LanguageSettings.shared.resolvedLanguage {
        case "pl":
            if n == 1 { return .one }
            let n10 = n % 10
            let n100 = n % 100
            if (2...4).contains(n10) && !(12...14).contains(n100) { return .few }
            return .many
        case "uk":
            let n10 = n % 10
            let n100 = n % 100
            if n10 == 1 && n100 != 11 { return .one }
            if (2...4).contains(n10) && !(12...14).contains(n100) { return .few }
            return .many
        case "cs":
            if n == 1 { return .one }
            if (2...4).contains(n) { return .few }
            return .other
        case "fr":
            return n <= 1 ? .one : .other
        default:
            return n == 1 ? .one : .other
        }
    }
}

File: windows/Gutenkern/L10n.cs
using System.Globalization;
using System.IO;
using System.Text.Json;

namespace Gutenkern;

internal readonly record struct AppLanguage(string Code, string NativeName);

internal readonly record struct CreditPart(string Text, Uri? Link);

internal static class L10n
{
    public static readonly AppLanguage[] Languages =
    [
        new("en", "English"),
        new("de", "Deutsch"),
        new("fr", "Français"),
        new("pl", "Polski"),
        new("cs", "Čeština"),
        new("es", "Español"),
        new("sv", "Svenska"),
        new("it", "Italiano"),
        new("uk", "Українська")
    ];

    private static readonly HashSet<string> Supported = ["en", "de", "fr", "pl", "cs", "es", "sv", "it", "uk"];
    private static readonly Dictionary<string, Dictionary<string, string>> Catalog = LoadCatalog();
    private static Dictionary<string, string> Table = [];
    private static string Language = "en";

    static L10n()
    {
        AppSettings.Load();
        ApplyPreference(AppSettings.Language);
    }

    public static string FieldWhat => T("fieldWhat");
    public static string FieldWhatHint => T("fieldWhatHint");
    public static string GroupCopied => T("groupCopied");
    public static string ClickHint => T("clickHint").Replace("{modifier}", "Ctrl");
    public static string NewUnkernedPairs => T("newUnkernedPairs");
    public static string Format => T("format");
    public static string FormatFontLab => T("formatFontLab");
    public static string FormatGlyphs => T("formatGlyphs");
    public static string Result => T("result");
    public static string ResultAs => T("resultAs");
    public static string ResultLayoutRow => T("resultLayoutRow");
    public static string ResultLayoutColumn => T("resultLayoutColumn");
    public static string Copy => T("copy");
    public static string Copied => T("copied");
    public static string CopyAll => T("copyAll");
    public static string File => T("file");
    public static string Save => T("save");
    public static string SaveEllipsis => T("saveEllipsis");
    public static string SaveAndOpenEllipsis => T("saveAndOpenEllipsis");
    public static string SaveAsFile => T("saveAsFile");
    public static string SaveFailed => T("saveFailed");
    public static string StrikeThrough => T("strikeThrough");
    public static string ClearStrike => T("clearStrike");
    public static string ResultPlaceholder => T("resultPlaceholder");
    public static string ChooseFormat => T("chooseFormat");
    public static string ResetProgress => T("resetProgress");
    public static string Undo => T("undo");
    public static string Redo => T("redo");
    public static string TextDocument => T("textDocument");
    public static string AllFiles => T("allFiles");
    public static string Settings => T("settings");
    public static string LanguageLabel => T("language");
    public static string LanguageSystem => T("languageSystem");
    public static string Groups => T("groups");
    public static string Plan => T("plan");
    public static string Help => T("help");
    public static string About => T("about");
    public static string AboutBody => T("aboutBody");
    public static string AboutContact => T("aboutContact");
    public static string AboutCopyright => T("aboutCopyright");
    public const string AuthorEmail = "armos1999@gmail.com";
    public const string AuthorWebsite = "arsenmosiichuk.in.ua";
    public const string VladWebsite = "zahrevsky.com";
    public const string OleksiiWebsite = "oleksii.shmalko.com";
    public const string SourceWebsite = "http://www.junikstudio.com/kernings/";
    public static Uri SourceWebsiteUrl { get; } = new(SourceWebsite);

    public static string AboutVersion(string version) => T("aboutVersion").Replace("{version}", version);

    public static List<CreditPart> AboutBodyParts() =>
        TemplateParts(AboutBody, new Dictionary<string, (string Label, Uri Url)>
        {
            [SourceWebsite] = (SourceWebsite, SourceWebsiteUrl)
        });

    public static List<CreditPart> AboutCreditsParts() =>
        TemplateParts(AboutCopyright, new Dictionary<string, (string Label, Uri Url)>
        {
            ["{arsenSite}"] = (AuthorWebsite, new Uri("https://" + AuthorWebsite)),
            ["{vladSite}"] = (VladWebsite, new Uri("https://" + VladWebsite)),
            ["{oleksiiSite}"] = (OleksiiWebsite, new Uri("https://" + OleksiiWebsite))
        });

    private static List<CreditPart> TemplateParts(
        string template,
        Dictionary<string, (string Label, Uri Url)> replacements)
    {
        var parts = new List<CreditPart>();
        var remaining = template;
        while (true)
        {
            string? nextKey = null;
            var nextIndex = -1;
            foreach (var key in replacements.Keys)
            {
                var index = remaining.IndexOf(key, StringComparison.Ordinal);
                if (index >= 0 && (nextIndex < 0 || index < nextIndex))
                {
                    nextKey = key;
                    nextIndex = index;
                }
            }

            if (nextKey is null)
            {
                break;
            }

            if (nextIndex > 0)
            {
                parts.Add(new CreditPart(remaining[..nextIndex], null));
            }

            var (label, url) = replacements[nextKey];
            parts.Add(new CreditPart(label, url));
            remaining = remaining[(nextIndex + nextKey.Length)..];
        }

        if (remaining.Length > 0)
        {
            parts.Add(new CreditPart(remaining, null));
        }

        return parts;
    }

    public static string GroupLabel(KerningGroup group)
    {
        var name = group switch
        {
            KerningGroup.Capitals => T("groupCapitals"),
            KerningGroup.SmallCaps => T("groupSmallCaps"),
            KerningGroup.Lowercase => T("groupLowercase"),
            KerningGroup.Punctuation => T("groupPunctuation"),
            KerningGroup.NonAlphabetic => T("groupNonAlphabetic"),
            KerningGroup.LiningFigures => T("groupLiningFigures"),
            KerningGroup.OldstyleFigures => T("groupOldstyleFigures"),
            _ => throw new ArgumentOutOfRangeException(nameof(group), group, null)
        };
        return name;
    }

    public static string GroupSidebarLabel(KerningGroup group, int done, int total)
    {
        var label = GroupLabel(group);
        if (total <= 0)
        {
            return label;
        }

        label += $" ⋅ {GroupedCount(done)}/{GroupedCount(total)}";
        if (done == total)
        {
            label += " ✓";
        }

        return label;
    }

    public static void ApplyPreference(string preference)
    {
        Language = Resolve(preference);
        Table = Merge(Language);
    }

    private static string T(string key) => Table.TryGetValue(key, out var value) ? value : key;

    private static string Resolve(string preference)
    {
        if (preference != AppSettings.SystemLanguage && Supported.Contains(preference))
        {
            return preference;
        }

        var culture = CultureInfo.CurrentUICulture;
        while (!string.IsNullOrEmpty(culture.Name) && !culture.Equals(CultureInfo.InvariantCulture))
        {
            var code = culture.TwoLetterISOLanguageName;
            if (Supported.Contains(code))
            {
                return code;
            }

            culture = culture.Parent;
        }

        return "en";
    }

    private static Dictionary<string, string> Merge(string language)
    {
        var merged = new Dictionary<string, string>();
        if (Catalog.TryGetValue("en", out var english) && english is not null)
        {
            foreach (var pair in english)
            {
                merged[pair.Key] = pair.Value;
            }
        }

        if (language != "en" && Catalog.TryGetValue(language, out var overlay) && overlay is not null)
        {
            foreach (var pair in overlay)
            {
                merged[pair.Key] = pair.Value;
            }
        }

        return merged;
    }

    private static Dictionary<string, Dictionary<string, string>> LoadCatalog()
    {
        try
        {
            var path = CatalogPath();
            if (path is null)
            {
                return [];
            }

            return JsonSerializer.Deserialize<Dictionary<string, Dictionary<string, string>>>(
                System.IO.File.ReadAllText(path)) ?? [];
        }
        catch
        {
            return [];
        }
    }

    private static string? CatalogPath()
    {
        var nextToExe = Path.Combine(AppContext.BaseDirectory, "l10n.json");
        if (System.IO.File.Exists(nextToExe))
        {
            return nextToExe;
        }

        var fromSource = Path.GetFullPath(Path.Combine(AppContext.BaseDirectory, "..", "..", "..", "..", "..", "core", "l10n.json"));
        return System.IO.File.Exists(fromSource) ? fromSource : null;
    }

    private enum PluralCategory
    {
        One,
        Few,
        Many,
        Other
    }

    public static string PairCount(int count)
    {
        var key = Plural(count) switch
        {
            PluralCategory.One => "pairOne",
            PluralCategory.Few => "pairFew",
            PluralCategory.Many => "pairMany",
            _ => "pairOther"
        };
        return T(key).Replace("{count}", GroupedCount(count));
    }

    public static string PairPercent(int percent) =>
        T("pairPercent").Replace("{percent}", percent.ToString(CultureInfo.InvariantCulture));

    public static string CopiedPairs(string pairs) =>
        T("copiedPairs").Replace("{pairs}", pairs);

    public static string PairProgress(int done, int total)
    {
        var progress = $"{GroupedCount(done)}/{GroupedCount(total)}";
        return PairCount(total).Replace(GroupedCount(total), progress);
    }

    public static int PairProgressPercent(int done, int total) =>
        total <= 0 ? 0 : (int)Math.Round(100.0 * done / total);

    public static string GroupedCount(int count)
    {
        var sign = count < 0 ? "-" : "";
        var digits = Math.Abs(count).ToString(CultureInfo.InvariantCulture);
        var grouped = new System.Text.StringBuilder();
        for (var index = 0; index < digits.Length; index++)
        {
            var fromEnd = digits.Length - index;
            if (index > 0 && fromEnd % 3 == 0)
            {
                grouped.Append('\u202F');
            }

            grouped.Append(digits[index]);
        }

        return sign + grouped.ToString();
    }

    private static PluralCategory Plural(int count)
    {
        var n = Math.Abs(count);
        switch (Language)
        {
            case "pl":
                if (n == 1)
                {
                    return PluralCategory.One;
                }

                if (n % 10 is >= 2 and <= 4 && n % 100 is not (>= 12 and <= 14))
                {
                    return PluralCategory.Few;
                }

                return PluralCategory.Many;
            case "uk":
                var n10 = n % 10;
                var n100 = n % 100;
                if (n10 == 1 && n100 != 11)
                {
                    return PluralCategory.One;
                }

                if (n10 is >= 2 and <= 4 && n100 is not (>= 12 and <= 14))
                {
                    return PluralCategory.Few;
                }

                return PluralCategory.Many;
            case "cs":
                if (n == 1)
                {
                    return PluralCategory.One;
                }

                return n is >= 2 and <= 4 ? PluralCategory.Few : PluralCategory.Other;
            case "fr":
                return n <= 1 ? PluralCategory.One : PluralCategory.Other;
            default:
                return n == 1 ? PluralCategory.One : PluralCategory.Other;
        }
    }
}
